From 2928ad06c78327559edd07b9685ca54c0ada4c20 Mon Sep 17 00:00:00 2001
From: Mohit Sabharwal <mohit@cloudera.com>
Date: Fri, 24 Feb 2017 14:59:03 -0500
Subject: [PATCH 0991/1164] CDH-48525 : Backport HIVE-15766 :
 DBNotificationlistener leaks
 JDOPersistenceManager : Use same
 PersistenceManager for metadata and notification

Note that this is an optimistic backport - +1'd but still waiting for commit upstream.

Change-Id: I4927b699f54d3fe855760ce836a395b121f8ca76
---
 .../hcatalog/listener/DbNotificationListener.java  |   30 +++++++-------------
 1 file changed, 11 insertions(+), 19 deletions(-)

diff --git a/hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java b/hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java
index e467b83..ffcd5fe 100644
--- a/hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java
+++ b/hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java
@@ -19,6 +19,7 @@
 
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.hive.metastore.HiveMetaStore;
 import org.apache.hadoop.hive.metastore.api.Function;
 import org.apache.hadoop.hive.metastore.api.Index;
 import org.apache.hadoop.hive.metastore.events.AddIndexEvent;
@@ -74,23 +75,16 @@
   // HiveConf rather than a Configuration.
   private HiveConf hiveConf;
   private MessageFactory msgFactory;
-  private RawStore rs;
-
-  private synchronized void init(HiveConf conf) {
-    try {
-      rs = RawStoreProxy.getProxy(conf, conf,
-          conf.getVar(HiveConf.ConfVars.METASTORE_RAW_STORE_IMPL), 999999);
-    } catch (MetaException e) {
-      LOG.error("Unable to connect to raw store, notifications will not be tracked", e);
-      rs = null;
-    }
-    if (cleaner == null && rs != null) {
-      cleaner = new CleanerThread(conf, rs);
+
+  private synchronized void init(HiveConf conf) throws MetaException {
+    if (cleaner == null) {
+      cleaner = new CleanerThread(conf, RawStoreProxy.getProxy(conf, conf,
+          conf.getVar(HiveConf.ConfVars.METASTORE_RAW_STORE_IMPL), 999999));
       cleaner.start();
     }
   }
 
-  public DbNotificationListener(Configuration config) {
+  public DbNotificationListener(Configuration config) throws MetaException {
     super(config);
     // The code in MetastoreUtils.getMetaStoreListeners() that calls this looks for a constructor
     // with a Configuration parameter, so we have to declare config as Configuration.  But it
@@ -330,12 +324,10 @@ private int now() {
   }
 
   private void enqueue(NotificationEvent event) {
-    if (rs != null) {
-      synchronized(NOTIFICATION_TBL_LOCK) {
-        rs.addNotificationEvent(event);
-      }
-    } else {
-      LOG.warn("Dropping event " + event + " since notification is not running.");
+    synchronized(NOTIFICATION_TBL_LOCK) {
+      LOG.debug("DbNotificationListener: Processing : " + event.getEventId() +
+          " : " + event.getMessage());
+      HiveMetaStore.HMSHandler.getRawStore().addNotificationEvent(event);
     }
   }
 
-- 
1.7.9.5

